### ISSUE TITLE
converting PCMSolver from MODULE to CONFIG

### ISSUE THREAD

**Issue by loriab** [PCMSolver/pcmsolver#32]
Hi @robertodr, if you're willing to do some CMake revamping for PCMSolver, that'd be a great help. Ask me any questions, no matter how small.

* Yes, for the most part if you copy [libefp](https://github.com/psi4/libefp/tree/config3) for simple, [gdma](https://github.com/psi4/gdma) for Fortran, and [CheMPS2](https://github.com/SebWouters/CheMPS2) for complex and in particular simultaneous shared/static build, that'll be a good pattern. Basically, you want to get an exported target PCMSolverConfig.cmake being installed with the library, then read-in-able ultimately [here](https://github.com/psi4/psi4/blob/master/psi4/CMakeLists.txt#L60-L64).

* In psi4, the `PCMSOLVER_ROOT` will go away and be replaced by `CMAKE_PREFIX_PATH`. The `find_package(` will go from MODULE to Config (FindPCMSolver.cmake goes away). [This](https://github.com/psi4/psi4/blob/master/external/pcmsolver/CMakeLists.txt) it looks like I already started adapting. It should look a lot like the others (except GTFock) in the externals folder.

* One change in the new Psi4 superbuild is that instead of source (C++ from objdir, Py from source) and installed (C++ from install_prefix; Py from install_prefix), we now have inplace (C++ from objdir, Py from source), staged (C++ from objdir, Py from objdir), and installed (C++ from install_prefix; Py from install_prefix). The inplace is a special case that's handy for developing Py-side, but we don't demand that it be all-powerful. In fact, since pcm had to do `.in` and checked-in `.in` before, I'd say just arrange so that it runs from staged and don't try for inplace.

* The main trick is going to be all of pcm's dependencies since ideally, there's no local absolute paths in the Config*cmake files. But I think pcm pretty much builds its own dependencies, so we can make that as difficult (and proper cmake) or as simple (build them all and install alongside the pcm library) as we choose.

Basically, let me know any questions and how deeply you want to get involved in dependency flexibility. Thank goodness you've only Boost header libs.

* At one point, I thought pcm was actually building a python module, and I fished up the below. take anything that's useful.

```
>>> cat cmake/custom/python_sitepackages.cmake 
# extracted from FindPythonLibsNew

#STATIC
#n module extension
#   execute_process(COMMAND "${PYTHON_EXECUTABLE}" -c
#                        "import distutils.sysconfig, sys; sys.stdout.write(distutils.sysconfig.get_config_var('SO'))"
#                   OUTPUT_VARIABLE PYTHON_MODULE_EXTENSION
#                   RESULT_VARIABLE _PYTHON_MODULE_EXTENSION)
#
#   # Get Python site-packages directory
#   execute_process(COMMAND "${PYTHON_EXECUTABLE}" -c
#                        "import distutils.sysconfig, sys; sys.stdout.write(distutils.sysconfig.get_python_lib())"
#                    OUTPUT_VARIABLE PYTHON_SITE_PACKAGES
#                    OUTPUT_STRIP_TRAILING_WHITESPACE)



# According to http://stackoverflow.com/questions/646518/python-how-to-detect-debug-interpreter
# testing whether sys has the gettotalrefcount function is a reliable, cross-platform
# way to detect a CPython debug interpreter.
#
# The library suffix is from the config var LDVERSION sometimes, otherwise
# VERSION. VERSION will typically be like "2.7" on unix, and "27" on windows.
execute_process(COMMAND "${PYTHON_EXECUTABLE}" "-c"
    "from distutils import sysconfig as s;import sys;import struct;
print('.'.join(str(v) for v in sys.version_info));
print(sys.prefix);
print(s.get_python_inc(plat_specific=True));
print(s.get_python_lib(plat_specific=True));
print(s.get_config_var('SO'));
print(hasattr(sys, 'gettotalrefcount')+0);
print(struct.calcsize('@P'));
print(s.get_config_var('LDVERSION') or s.get_config_var('VERSION'));
"
    RESULT_VARIABLE _PYTHON_SUCCESS
    OUTPUT_VARIABLE _PYTHON_VALUES
    ERROR_VARIABLE _PYTHON_ERROR_VALUE
    OUTPUT_STRIP_TRAILING_WHITESPACE)

if(NOT _PYTHON_SUCCESS MATCHES 0)
    if(PythonLibsNew_FIND_REQUIRED)
        message(FATAL_ERROR
            "Python config failure:\n${_PYTHON_ERROR_VALUE}")
    endif()
    set(PYTHONLIBS_FOUND FALSE)
    return()
endif()

# Convert the process output into a list
string(REGEX REPLACE ";" "\\\\;" _PYTHON_VALUES ${_PYTHON_VALUES})
string(REGEX REPLACE "\n" ";" _PYTHON_VALUES ${_PYTHON_VALUES})
list(GET _PYTHON_VALUES 0 _PYTHON_VERSION_LIST)
list(GET _PYTHON_VALUES 1 PYTHON_PREFIX)
list(GET _PYTHON_VALUES 2 PYTHON_INCLUDE_DIR)
list(GET _PYTHON_VALUES 3 PYTHON_SITE_PACKAGES)
list(GET _PYTHON_VALUES 4 PYTHON_MODULE_EXTENSION)
list(GET _PYTHON_VALUES 5 PYTHON_IS_DEBUG)
list(GET _PYTHON_VALUES 6 PYTHON_SIZEOF_VOID_P)
list(GET _PYTHON_VALUES 7 PYTHON_LIBRARY_SUFFIX)

# Make sure the Python has the same pointer-size as the chosen compiler
# Skip if CMAKE_SIZEOF_VOID_P is not defined
if(CMAKE_SIZEOF_VOID_P AND (NOT "${PYTHON_SIZEOF_VOID_P}" STREQUAL "${CMAKE_SIZEOF_VOID_P}"))
    if(PythonLibsNew_FIND_REQUIRED)
        math(EXPR _PYTHON_BITS "${PYTHON_SIZEOF_VOID_P} * 8")
        math(EXPR _CMAKE_BITS "${CMAKE_SIZEOF_VOID_P} * 8")
        message(FATAL_ERROR
            "Python config failure: Python is ${_PYTHON_BITS}-bit, "
            "chosen compiler is  ${_CMAKE_BITS}-bit")
    endif()
    set(PYTHONLIBS_FOUND FALSE)
    return()
endif()

# The built-in FindPython didn't always give the version numbers
string(REGEX REPLACE "\\." ";" _PYTHON_VERSION_LIST ${_PYTHON_VERSION_LIST})
list(GET _PYTHON_VERSION_LIST 0 PYTHON_VERSION_MAJOR)
list(GET _PYTHON_VERSION_LIST 1 PYTHON_VERSION_MINOR)
list(GET _PYTHON_VERSION_LIST 2 PYTHON_VERSION_PATCH)

# Make sure all directory separators are '/'
string(REGEX REPLACE "\\\\" "/" PYTHON_PREFIX ${PYTHON_PREFIX})
string(REGEX REPLACE "\\\\" "/" PYTHON_INCLUDE_DIR ${PYTHON_INCLUDE_DIR})
string(REGEX REPLACE "\\\\" "/" PYTHON_SITE_PACKAGES ${PYTHON_SITE_PACKAGES})
```

**Comment by robertodr** [comment]
@loriab thanks for the very detailed PR. I'll be working on it in the next days. If you have an active fork, I'd suggest to nuke it and re-fork. The `master` branch was subjected to my iconoclastic, history-rewriting fury. The `release/1.Y` is fine, but I'd prefer PR to be directed to `master`

**Comment by robertodr** [comment]
Closed by #38 @loriab open a new issue for anything that's missing PCMSolver-side to be fully Psi4-friendly (I am thinking the `__init__.py` discussion we had)
